### ISSUE TITLE
Wanted basic features

### ISSUE THREAD

**Issue by ice1000** [owo-lang/rfcs#2]

**Comment by ice1000** [comment]
+ GADT (Haskell, Agda, Idris) and Generalized Algebraic Codata Type (Agda)
+ Dependent (Co)pattern matching (Agda)

**Comment by ice1000** [comment]
+ Reflection (Agda, Idris)

**Comment by ice1000** [comment]
+ Prop, Proof Irrelevance (Agda, Idris, Coq)

**Comment by ice1000** [comment]
+ Sized Type (?) (Agda)

owo-lang/OwO#5 Redy!

**Comment by ice1000** [comment]
+ Termination check, possibly cross function? (Agda, Idris, Coq)

**Comment by ice1000** [comment]
+ Module System (Agda, OCaml)
  + import
  + open
  + `open {{ ... }}`
  + parameterization
  + hiding
  + using
  + renaming

**Comment by ice1000** [comment]
+ Automatic Generalization (Haskell, Idris, OCaml)

**Comment by ice1000** [comment]
+ Pattern Matching without K (Agda)

**Comment by ice1000** [comment]
+ Overloading over lambdas (Agda)

Cubical Type Theories?

**Comment by ice1000** [comment]
+ Foreign Function Interface (*)

**Comment by ice1000** [comment]
+ Instance Arguments (Agda)

I don't want typeclass/type instance as language component.

**Comment by ice1000** [comment]
+ Strict evaluation for data, Lazy evaluation for codata? (^)

**Comment by ice1000** [comment]
+ Global overloading (Idris)
+ Local shadowing (^Agda)

**Comment by ice1000** [comment]
+ Pattern Matching without K

**Comment by ice1000** [comment]
+ Metavariables (Haskell, Agda, Idris)

**Comment by ice1000** [comment]
+ Universal Polymorphism (Agda, Idris, Coq)

**Comment by anqur** [comment]
What about n-truncated types like HoTT-Agda's to introduce `hProp`/`hSet`, make identity types and homotopy equivalence *equivalent*, and axiom K a theorem for `hSet`.

**Comment by ice1000** [comment]
> What about n-truncated types like HoTT-Agda's to introduce `hProp`/`hSet`, make identity types and homotopy equivalence _equivalent_, and axiom K a theorem for `hSet`.

It has already been covered as `Prop` and `Proof Irrelevance`

**Comment by ice1000** [comment]
+ Types with boolean refinement (F*), may replace sized type?

**Comment by anqur** [comment]
It's an urge to give a research about the ability of modeling HoTT by UTT, since the base standard library of OwO should start with things like in HoTT-Agda's n-truncated types, since then we can easily model h-props and h-sets by postulating different axioms, and even forget about UA when modelling all the groupoids.

**Comment by anqur** [comment]
And I will leave a comment about the design of `record`-like stuff, since I think Agda's (co-)inductive records are weird to construct...

**Comment by ice1000** [comment]
Yes, forget about record constructors. We'll use copatterns to replace record constructors!

**Comment by ice1000** [comment]
After a little study about cutt, I'd say proof irrelevance is no longer useful when we have path types.
